File: audio.py
# -*- coding: utf-8 -*-
"""
audio.py
========
Kleine Sound-Engine für die Spielesammlung.

- Die Effekte werden zur Laufzeit synthetisiert (kurze Töne/Rauschen), damit
  KEINE externen WAV-Dateien nötig sind und auch KEIN numpy gebraucht wird:
  wir bauen rohe 16-Bit-PCM-Samples mit dem 'array'-Modul und übergeben sie
  direkt an pygame.mixer.Sound(buffer=...).
- play(name, settings) spielt einen Effekt nur, wenn settings["sound"] aktiv ist,
  und nutzt settings["volume"] als Lautstärke.
- rumble(settings, ms) löst Gamepad-Vibration aus, sofern ein Controller
  vorhanden und settings["haptik"] aktiv ist (sonst wirkungslos).

Alles ist mit try/except abgesichert: fehlt Audio-Hardware oder Mixer, läuft
das Spiel trotzdem (nur eben ohne Ton).
"""

import math
import random
from array import array

import pygame

_available = False          # ist der Mixer nutzbar?
_cache = {}                 # name -> pygame.mixer.Sound
_tone_cache = {}            # (freq, dur, wave) -> pygame.mixer.Sound (frei erzeugt)
_joysticks = []             # initialisierte Gamepads (für Rumble)

# Spezifikation der Effekte. f0 = Startfrequenz, f1 = Zielfrequenz (Sweep),
# dur = Dauer in Sekunden, wave = Wellenform, vol = relative Amplitude.
_SPECS = {
    "click":    dict(f0=880, dur=0.05, wave="square", vol=0.25),
    "select":   dict(f0=700, dur=0.05, wave="sine",   vol=0.25),
    "eat":      dict(f0=660, f1=1320, dur=0.09, wave="square", vol=0.30),
    "bounce":   dict(f0=440, dur=0.05, wave="square", vol=0.30),
    "point":    dict(f0=520, f1=1040, dur=0.18, wave="square", vol=0.35),
    "shoot":    dict(f0=900, f1=300,  dur=0.12, wave="saw",    vol=0.28),
    "explode":  dict(f0=200, dur=0.22, wave="noise",  vol=0.40),
    "hit":      dict(f0=160, dur=0.18, wave="square", vol=0.40),
    "rotate":   dict(f0=600, dur=0.04, wave="sine",   vol=0.22),
    "lock":     dict(f0=200, dur=0.07, wave="square", vol=0.30),
    "line":     dict(f0=400, f1=1200, dur=0.25, wave="square", vol=0.40),
    "merge":    dict(f0=500, f1=760,  dur=0.10, wave="sine",   vol=0.30),
    "move":     dict(f0=330, dur=0.03, wave="square", vol=0.18),
    "gameover": dict(f0=440, f1=120,  dur=0.50, wave="saw",    vol=0.40),
    "win":      dict(f0=520, f1=1300, dur=0.45, wave="square", vol=0.40),
    "powerup":  dict(f0=600, f1=1200, dur=0.18, wave="square", vol=0.32),
    "level":    dict(f0=440, f1=1100, dur=0.32, wave="square", vol=0.38),
}


def init():
    """Initialisiert Mixer (falls nötig) und erzeugt alle Effekte vorab."""
    global _available
    try:
        if pygame.mixer.get_init() is None:
            pygame.mixer.init(frequency=44100, size=-16, channels=1)
        pygame.mixer.set_num_channels(16)
        _available = pygame.mixer.get_init() is not None
    except Exception:
        _available = False
        return

    # Gamepads für Rumble vorbereiten (optional).
    try:
        pygame.joystick.init()
        for i in range(pygame.joystick.get_count()):
            js = pygame.joystick.Joystick(i)
            js.init()
            _joysticks.append(js)
    except Exception:
        pass

    for name in _SPECS:
        try:
            _cache[name] = _build(_SPECS[name])
        except Exception:
            _cache[name] = None


def _build(spec):
    """Synthetisiert einen Effekt und gibt ein pygame.mixer.Sound zurück."""
    freq_hz, fmt, channels = pygame.mixer.get_init()
    n = max(1, int(freq_hz * spec["dur"]))
    f0 = spec["f0"]
    f1 = spec.get("f1", f0)
    wave = spec["wave"]
    amp = spec["vol"] * 32767

    att = max(1, int(n * 0.05))          # kurzer Ein-/Ausblendbereich gegen Knackser
    rel = max(1, int(n * 0.20))

    buf = array("h")
    phase = 0.0
    for i in range(n):
        frac = i / n
        f = f0 + (f1 - f0) * frac
        phase += f / freq_hz             # Phasenakkumulation (für Sweeps korrekt)
        p = phase % 1.0

        if wave == "sine":
            val = math.sin(2 * math.pi * p)
        elif wave == "square":
            val = 1.0 if p < 0.5 else -1.0
        elif wave == "saw":
            val = 2.0 * p - 1.0
        else:  # noise
            val = random.uniform(-1.0, 1.0)

        # Hüllkurve: einblenden, ausklingen.
        env = min(1.0, i / att, (n - i) / rel)
        s = int(max(-1.0, min(1.0, val)) * amp * env)

        buf.append(s)
        if channels == 2:                # bei Stereo denselben Wert auf L und R
            buf.append(s)

    return pygame.mixer.Sound(buffer=buf.tobytes())


def play(name, settings=None):
    """Spielt Effekt 'name', falls Sound aktiv ist."""
    if not _available:
        return
    if settings is not None and not settings.get("sound", True):
        return
    snd = _cache.get(name)
    if snd is None:
        return
    try:
        vol = 0.6 if settings is None else float(settings.get("volume", 0.6))
        snd.set_volume(max(0.0, min(1.0, vol)))
        snd.play()
    except Exception:
        pass


def tone(freq, dur=0.18, settings=None, wave="sine", vol=0.35):
    """Spielt einen frei wählbaren Ton (Frequenz in Hz).

    Anders als play() sind die Effekte nicht vorab definiert: der Ton wird bei
    Bedarf synthetisiert und danach gecacht (nach Frequenz/Dauer/Wellenform).
    Gedacht z.B. für Simon/Senso, das je Feld einen eigenen Klang braucht.
    Respektiert - wie play() - settings["sound"] und settings["volume"].
    """
    if not _available:
        return
    if settings is not None and not settings.get("sound", True):
        return
    key = (round(float(freq), 1), round(float(dur), 3), wave)
    snd = _tone_cache.get(key)
    if snd is None:
        try:
            snd = _build(dict(f0=freq, dur=dur, wave=wave, vol=1.0))
        except Exception:
            snd = None
        _tone_cache[key] = snd
    if snd is None:
        return
    try:
        v = 0.6 if settings is None else float(settings.get("volume", 0.6))
        snd.set_volume(max(0.0, min(1.0, v * vol)))
        snd.play()
    except Exception:
        pass


def rumble(settings=None, ms=120, strong=0.6, weak=0.4):
    """Löst Gamepad-Vibration aus, falls Haptik aktiv und Controller vorhanden."""
    if settings is not None and not settings.get("haptik", False):
        return
    for js in _joysticks:
        try:
            js.rumble(strong, weak, ms)     # verfügbar ab pygame 2 / SDL2
        except Exception:
            pass


# ---------------------------------------------------------------------------
#  Musik-Schleifen (z.B. Level-Soundtrack in Geometry Dash)
# ---------------------------------------------------------------------------
#
# Die Musik besteht aus mehreren STIMMEN (Drums, Bass, Melodie ...), jede ein
# eigener, gleich langer Sound, der auf einem eigenen reservierten Kanal in
# Dauerschleife läuft. So muss Python nichts mischen - das macht SDL.
#
# Eine Stimme ist monophon: sie wird aus Noten zusammengesetzt, deren Samples
# EINMAL synthetisiert (gecacht) und dann per Byte-Slice an ihre Position
# kopiert werden. Das kostet für eine 30-Sekunden-Schleife nur Millisekunden -
# Sample für Sample (wie bei den Effekten) wäre es um Sekunden zu langsam.
#
# Die Musik gehört einem Besitzer (dem Spiel-Objekt). main.py ruft jedes Frame
# music_tick(aktives_spiel) auf: ist der Besitzer nicht mehr aktiv, endet die
# Musik; ist er pausiert, pausiert sie mit.

MUSIC_CHANNELS = 4          # reservierte Kanäle 0..3 (Sound.play() nimmt sie nie)
_music = {"owner": None, "channels": [], "sounds": [], "vols": [],
          "paused": False, "volume": 1.0, "settings": None}
_note_cache = {}            # (freq, samples, wave, vol, decay, rate) -> bytes (mono)


def _reserve_music_channels():
    """Reserviert die Musik-Kanäle (einmalig, nach dem Mixer-Start)."""
    if _music["channels"] or not _available:
        return bool(_music["channels"])
    try:
        if pygame.mixer.get_num_channels() < 16 + MUSIC_CHANNELS:
            pygame.mixer.set_num_channels(16 + MUSIC_CHANNELS)
        pygame.mixer.set_reserved(MUSIC_CHANNELS)
        _music["channels"] = [pygame.mixer.Channel(i) for i in range(MUSIC_CHANNELS)]
    except Exception:
        _music["channels"] = []
    return bool(_music["channels"])


def mixer_rate():
    """Abtastrate des Mixers in Hz (44100, falls kein Mixer läuft)."""
    try:
        init = pygame.mixer.get_init()
        return init[0] if init else 44100
    except Exception:
        return 44100


def note_bytes(freq, dur, wave="square", vol=0.3, decay=0.0, rate=None):
    """Mono-16-Bit-Samples EINER Note als bytes (gecacht).

    freq  : Frequenz in Hz (bei kick/snare/hat ohne Bedeutung)
    dur   : Länge in Sekunden
    wave  : sine / square / saw / triangle / noise / kick / snare / hat
    vol   : Lautstärke 0..1
    decay : 0 = gleichmäßig mit kurzem Ausklang, > 0 = exponentielles
            Abklingen (Faktor pro Sekunde, z.B. 8 für gezupfte Töne)
    """
    rate = rate or mixer_rate()
    n = max(1, int(rate * dur))
    key = (round(float(freq), 2), n, wave, round(float(vol), 3),
           round(float(decay), 2), rate)
    data = _note_cache.get(key)
    if data is not None:
        return data
    buf = array("h")
    amp = max(0.0, min(1.0, vol)) * 32767
    rng = random.Random(int(freq * 100) + n)   # Rauschen reproduzierbar
    att = max(1, int(rate * 0.004))
    rel = max(1, min(n // 3, int(rate * 0.03)))
    phase = 0.0
    step = freq / rate
    for i in range(n):
        tt = i / rate
        if wave == "kick":
            phase += (150.0 * math.exp(-tt * 28.0) + 45.0) / rate
            val = math.sin(2 * math.pi * phase)
            env = math.exp(-tt * 9.0)
        elif wave == "snare":
            phase += 190.0 / rate
            val = 0.65 * rng.uniform(-1.0, 1.0) + 0.35 * math.sin(2 * math.pi * phase)
            env = math.exp(-tt * 18.0)
        elif wave == "hat":
            val = rng.uniform(-1.0, 1.0)
            env = math.exp(-tt * 45.0)
        else:
            phase += step
            p = phase % 1.0
            if wave == "sine":
                val = math.sin(2 * math.pi * p)
            elif wave == "saw":
                val = 2.0 * p - 1.0
            elif wave == "triangle":
                val = 4.0 * p - 1.0 if p < 0.5 else 3.0 - 4.0 * p
            elif wave == "noise":
                val = rng.uniform(-1.0, 1.0)
            else:
                val = 1.0 if p < 0.5 else -1.0
            env = math.exp(-tt * decay) if decay > 0 else 1.0
        env *= min(1.0, i / att, (n - i) / rel)
        buf.append(int(max(-1.0, min(1.0, val)) * amp * env))
    data = buf.tobytes()
    _note_cache[key] = data
    return data


def render_voice(notes, length, rate=None):
    """Baut eine Stimme: notes = [(start_sekunde, bytes), ...] -> Mono-bytes.

    'length' ist die Gesamtlänge in Sekunden. Überlappt eine Note die nächste
    oder das Ende, wird sie abgeschnitten (eine Stimme ist monophon).
    """
    rate = rate or mixer_rate()
    total = int(rate * length) * 2
    out = bytearray(total)
    for start, data in notes:
        pos = int(rate * start) * 2
        if pos < 0 or pos >= total:
            continue
        chunk = data[: total - pos]
        out[pos:pos + len(chunk)] = chunk
    return bytes(out)


def _mono_to_mixer(mono):
    """Mono-16-Bit-bytes -> Format des Mixers (bei Stereo: links = rechts)."""
    try:
        channels = pygame.mixer.get_init()[2]
    except Exception:
        channels = 2
    if channels == 1:
        return mono
    n = len(mono) // 2
    lo, hi = mono[0:n * 2:2], mono[1:n * 2:2]
    out = bytearray(n * 2 * channels)
    for c in range(channels):
        out[c * 2::channels * 2] = lo
        out[c * 2 + 1::channels * 2] = hi
    return bytes(out)


def music_play(voices, owner, settings=None, volume=1.0):
    """Startet eine Musik-Schleife aus mehreren Stimmen (max. MUSIC_CHANNELS).

    voices : Liste von (mono_bytes, relative_lautstärke) - alle gleich lang
    owner  : das Spiel, dem die Musik gehört (siehe music_tick)
    Eine bereits laufende Musik wird vorher beendet. Gibt True bei Erfolg.
    """
    music_stop()
    if not _available or not _reserve_music_channels():
        return False
    try:
        if abs(pygame.mixer.get_init()[1]) != 16:
            return False
    except Exception:
        return False
    sounds, vols = [], []
    for mono, vol in list(voices)[:MUSIC_CHANNELS]:
        try:
            sounds.append(pygame.mixer.Sound(buffer=_mono_to_mixer(mono)))
            vols.append(float(vol))
        except Exception:
            pass
    _music.update(owner=owner, sounds=sounds, vols=vols, paused=False,
                  volume=float(volume), settings=settings)
    _apply_music_volume()
    for ch, snd in zip(_music["channels"], sounds):
        try:
            ch.play(snd, loops=-1)
        except Exception:
            pass
    return bool(sounds)


def music_stop(owner=None):
    """Beendet die Musik (nur die von 'owner', falls angegeben)."""
    if owner is not None and _music["owner"] is not owner:
        return
    for ch in _music["channels"]:
        try:
            ch.stop()
        except Exception:
            pass
    _music.update(owner=None, sounds=[], vols=[], paused=False)


def music_set_volume(volume):
    """Relative Lautstärke der laufenden Musik (0..1, z.B. fürs Ausblenden)."""
    _music["volume"] = max(0.0, min(1.0, float(volume)))
    _apply_music_volume()


def music_playing(owner=None):
    """True, wenn gerade Musik läuft (optional: die von 'owner')."""
    return bool(_music["sounds"]) and (owner is None or _music["owner"] is owner)


def _apply_music_volume():
    settings = _music["settings"]
    base = 0.6
    if settings is not None:
        base = float(settings.get("volume", 0.6)) if settings.get("sound", True) else 0.0
    for ch, vol in zip(_music["channels"], _music["vols"]):
        try:
            ch.set_volume(max(0.0, min(1.0, base * vol * _music["volume"])))
        except Exception:
            pass


def music_tick(current):
    """Wird von main.py jedes Frame aufgerufen: die Musik folgt ihrem Besitzer.

    Ist 'current' nicht der Besitzer, endet die Musik. Pausiert der Besitzer
    (ESC), pausiert die Musik; Sound aus / Lautstärke wirken sofort.
    """
    owner = _music["owner"]
    if owner is None:
        return
    if current is not owner:
        music_stop()
        return
    want_pause = bool(getattr(owner, "paused", False))
    if want_pause != _music["paused"]:
        _music["paused"] = want_pause
        for ch in _music["channels"]:
            try:
                if want_pause:
                    ch.pause()
                else:
                    ch.unpause()
            except Exception:
                pass
    _apply_music_volume()

File: web/js/core/audio.js
/*
 * audio.js - synthetisierte Soundeffekte (Nachbau von audio.py)
 * ==============================================================
 * Keine Audiodateien: die Effekte werden beim ersten Nutzer-Klick/Tastendruck
 * per WebAudio aus denselben Spezifikationen wie in audio.py erzeugt.
 *
 *   PG.audio.play("eat");                  // respektiert Sound an/aus + Lautstärke
 *   PG.audio.tone(440, 0.18, "sine", 0.35); // freier Ton (z.B. Simon)
 *   PG.audio.rumble(120);                  // Gamepad-Vibration (falls vorhanden)
 */
(function () {
  "use strict";

  const PG = window.PG;

  const SPECS = {
    click: { f0: 880, dur: 0.05, wave: "square", vol: 0.25 },
    select: { f0: 700, dur: 0.05, wave: "sine", vol: 0.25 },
    eat: { f0: 660, f1: 1320, dur: 0.09, wave: "square", vol: 0.3 },
    bounce: { f0: 440, dur: 0.05, wave: "square", vol: 0.3 },
    point: { f0: 520, f1: 1040, dur: 0.18, wave: "square", vol: 0.35 },
    shoot: { f0: 900, f1: 300, dur: 0.12, wave: "saw", vol: 0.28 },
    explode: { f0: 200, dur: 0.22, wave: "noise", vol: 0.4 },
    hit: { f0: 160, dur: 0.18, wave: "square", vol: 0.4 },
    rotate: { f0: 600, dur: 0.04, wave: "sine", vol: 0.22 },
    lock: { f0: 200, dur: 0.07, wave: "square", vol: 0.3 },
    line: { f0: 400, f1: 1200, dur: 0.25, wave: "square", vol: 0.4 },
    merge: { f0: 500, f1: 760, dur: 0.1, wave: "sine", vol: 0.3 },
    move: { f0: 330, dur: 0.03, wave: "square", vol: 0.18 },
    gameover: { f0: 440, f1: 120, dur: 0.5, wave: "saw", vol: 0.4 },
    win: { f0: 520, f1: 1300, dur: 0.45, wave: "square", vol: 0.4 },
    powerup: { f0: 600, f1: 1200, dur: 0.18, wave: "square", vol: 0.32 },
    level: { f0: 440, f1: 1100, dur: 0.32, wave: "square", vol: 0.38 },
  };

  let ctx = null;
  let master = null;
  const cache = new Map();
  const toneCache = new Map();
  const noteCache = new Map();
  let music = null;

  function ensure() {
    if (ctx) return ctx;
    const AC = window.AudioContext || window.webkitAudioContext;
    if (!AC) return null;
    try {
      ctx = new AC();
      master = ctx.createGain();
      master.gain.value = 0.9;
      master.connect(ctx.destination);
    } catch (e) {
      ctx = null;
    }
    return ctx;
  }

  function build(spec) {
    const rate = ctx.sampleRate;
    const n = Math.max(1, Math.floor(rate * spec.dur));
    const buf = ctx.createBuffer(1, n, rate);
    const data = buf.getChannelData(0);
    const f0 = spec.f0, f1 = spec.f1 != null ? spec.f1 : spec.f0;
    const att = Math.max(1, Math.floor(n * 0.05));
    const rel = Math.max(1, Math.floor(n * 0.2));
    let phase = 0;
    for (let i = 0; i < n; i++) {
      const f = f0 + (f1 - f0) * (i / n);
      phase += f / rate;
      const p = phase % 1;
      let v;
      if (spec.wave === "sine") v = Math.sin(PG.TAU * p);
      else if (spec.wave === "square") v = p < 0.5 ? 1 : -1;
      else if (spec.wave === "saw") v = 2 * p - 1;
      else v = Math.random() * 2 - 1;
      const env = Math.min(1, i / att, (n - i) / rel);
      data[i] = v * spec.vol * env;
    }
    return buf;
  }

  function enabled() {
    return !!PG.settings.data.sound;
  }

  function playBuffer(buf, vol) {
    try {
      const src = ctx.createBufferSource();
      src.buffer = buf;
      const g = ctx.createGain();
      g.gain.value = Math.max(0, Math.min(1, vol));
      src.connect(g);
      g.connect(master);
      src.start();
    } catch (e) {}
  }

  PG.audio = {
    /** Beim ersten Nutzer-Ereignis aufrufen (Browser-Autoplay-Regeln). */
    unlock() {
      if (!ensure()) return;
      if (ctx.state === "suspended") ctx.resume().catch(() => {});
    },
    play(name) {
      if (!enabled() || !ensure() || ctx.state !== "running") return;
      const spec = SPECS[name];
      if (!spec) return;
      let buf = cache.get(name);
      if (!buf) {
        buf = build(spec);
        cache.set(name, buf);
      }
      playBuffer(buf, Number(PG.settings.data.volume));
    },
    tone(freq, dur = 0.18, wave = "sine", vol = 0.35) {
      if (!enabled() || !ensure() || ctx.state !== "running") return;
      const key = Math.round(freq * 10) + "|" + Math.round(dur * 1000) + "|" + wave;
      let buf = toneCache.get(key);
      if (!buf) {
        buf = build({ f0: freq, dur, wave, vol: 1 });
        if (toneCache.size > 200) toneCache.clear();
        toneCache.set(key, buf);
      }
      playBuffer(buf, Number(PG.settings.data.volume) * vol);
    },
    rumble(ms = 120, strong = 0.6, weak = 0.4) {
      if (!PG.settings.data.haptik || !navigator.getGamepads) return;
      for (const gp of navigator.getGamepads()) {
        const act = gp && gp.vibrationActuator;
        if (act && act.playEffect) {
          act.playEffect("dual-rumble", { duration: ms, strongMagnitude: strong, weakMagnitude: weak }).catch(() => {});
        }
      }
    },
    SPECS,

    // ----- Musik-Schleifen (Gegenstück zu music_play/music_tick in audio.py) ---
    // Eine Stimme ist eine Float32Array (mono, Abtastrate = musicRate()), aus
    // Noten zusammengesetzt (noteSamples + renderVoice). Alle Stimmen laufen als
    // AudioBufferSource mit loop=true synchron. Die Musik gehört einem Spiel:
    // app.js ruft jedes Frame musicTick(current) auf.
    musicRate() {
      return ensure() ? ctx.sampleRate : 44100;
    },
    /** Samples einer Note (gecacht), Parameter wie note_bytes() in audio.py. */
    noteSamples(freq, dur, wave = "square", vol = 0.3, decay = 0) {
      const rate = this.musicRate();
      const n = Math.max(1, Math.floor(rate * dur));
      const key = [Math.round(freq * 100), n, wave, Math.round(vol * 1000), Math.round(decay * 100), rate].join("|");
      let data = noteCache.get(key);
      if (data) return data;
      data = new Float32Array(n);
      const att = Math.max(1, Math.floor(rate * 0.004));
      const rel = Math.max(1, Math.min(Math.floor(n / 3), Math.floor(rate * 0.03)));
      let phase = 0;
      let seed = (Math.floor(freq * 100) + n) >>> 0;
      const noise = () => {
        seed = (Math.imul(seed, 1664525) + 1013904223) >>> 0;
        return (seed / 4294967296) * 2 - 1;
      };
      for (let i = 0; i < n; i++) {
        const tt = i / rate;
        let v, env;
        if (wave === "kick") {
          phase += (150 * Math.exp(-tt * 28) + 45) / rate;
          v = Math.sin(PG.TAU * phase);
          env = Math.exp(-tt * 9);
        } else if (wave === "snare") {
          phase += 190 / rate;
          v = 0.65 * noise() + 0.35 * Math.sin(PG.TAU * phase);
          env = Math.exp(-tt * 18);
        } else if (wave === "hat") {
          v = noise();
          env = Math.exp(-tt * 45);
        } else {
          phase += freq / rate;
          const p = phase % 1;
          if (wave === "sine") v = Math.sin(PG.TAU * p);
          else if (wave === "saw") v = 2 * p - 1;
          else if (wave === "triangle") v = p < 0.5 ? 4 * p - 1 : 3 - 4 * p;
          else if (wave === "noise") v = noise();
          else v = p < 0.5 ? 1 : -1;
          env = decay > 0 ? Math.exp(-tt * decay) : 1;
        }
        env *= Math.min(1, i / att, (n - i) / rel);
        data[i] = Math.max(-1, Math.min(1, v)) * vol * env;
      }
      if (noteCache.size > 400) noteCache.clear();
      noteCache.set(key, data);
      return data;
    },
    /** Stimme aus [[startSekunde, Float32Array], ...] mit Gesamtlänge in Sekunden. */
    renderVoice(notes, length) {
      const rate = this.musicRate();
      const total = Math.floor(rate * length);
      const out = new Float32Array(total);
      for (const [start, data] of notes) {
        const pos = Math.floor(rate * start);
        if (pos < 0 || pos >= total) continue;
        out.set(total - pos < data.length ? data.subarray(0, total - pos) : data, pos);
      }
      return out;
    },
    /** Startet Musik: voices = [[Float32Array, relVol], ...], owner = Spiel-Objekt. */
    music(voices, owner, volume = 1) {
      this.stopMusic();
      if (!ensure()) return false;
      const rate = ctx.sampleRate;
      const gain = ctx.createGain();
      gain.connect(master);
      const nodes = [];
      const start = ctx.currentTime + 0.05;
      for (const [data, vol] of voices) {
        try {
          const buf = ctx.createBuffer(1, data.length, rate);
          buf.getChannelData(0).set(data);
          const src = ctx.createBufferSource();
          src.buffer = buf;
          src.loop = true;
          const g = ctx.createGain();
          g.gain.value = vol;
          src.connect(g);
          g.connect(gain);
          src.start(start);
          nodes.push(src);
        } catch (e) {}
      }
      music = { owner, gain, nodes, volume, paused: false };
      this.applyMusicVolume();
      return nodes.length > 0;
    },
    stopMusic(owner) {
      if (!music || (owner && music.owner !== owner)) return;
      for (const n of music.nodes) {
        try {
          n.stop();
        } catch (e) {}
      }
      try {
        music.gain.disconnect();
      } catch (e) {}
      music = null;
    },
    setMusicVolume(v) {
      if (!music) return;
      music.volume = Math.max(0, Math.min(1, v));
      this.applyMusicVolume();
    },
    musicPlaying(owner) {
      return !!music && (!owner || music.owner === owner);
    },
    applyMusicVolume() {
      if (!music) return;
      const base = enabled() ? Number(PG.settings.data.volume) : 0;
      const target = music.paused ? 0 : base * music.volume;
      try {
        music.gain.gain.setTargetAtTime(target, ctx.currentTime, 0.02);
      } catch (e) {
        music.gain.gain.value = target;
      }
    },
    /** Jedes Frame aus app.js: Musik endet mit ihrem Spiel und pausiert mit ihm. */
    musicTick(current) {
      if (!music) return;
      if (current !== music.owner) {
        this.stopMusic();
        return;
      }
      music.paused = !!current.paused;
      this.applyMusicVolume();
    },
  };
})();
